### ISSUE TITLE
Oxigraph servers hangs with FILTER function in sparql

### ISSUE THREAD

**Issue by YoelisA** [oxigraph/oxigraph#215]
Hi,

We deployed an oxigraph server and it appears the server hangs until it crashes when we use a FILTER function in the sparql query.
[When the predicate is set to true, it returns the data as expected](https://rcf-sparql.demo.logilab.fr/#query=PREFIX%20rcf%3A%20%3Chttp%3A%2F%2Frcf.logilab.fr%2F%3E%0APREFIX%20rdfs%3A%20%3Chttp%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23%3E%0APREFIX%20xsd%3A%20%3Chttp%3A%2F%2Fwww.w3.org%2F2001%2FXMLSchema%23%3E%0ASELECT%20DISTINCT%20*%0A%20%20%20%20%20%20WHERE%20%7B%0A%20%20%09%09%3FexpenseType%20rdfs%3AsubClassOf%20rcf%3AD%C3%A9penseImprim%C3%A9e.%0A%20%20%20%20%20%20%20%20%3FexpenseType%20rdfs%3Alabel%20%3Flabel.%0A%20%20%09%09%3Fexpense%20a%20%3FexpenseType.%0A%20%20%09%09%3Fjournee%20rcf%3AaPourDepense%20%3Fexpense.%0A%20%20%09%09%3Fjournee%20rcf%3AaPourDate%20%3Fdate.%20%0A%20%20%09%09FILTER%20(true)%0A%20%20%20%20%20%20%7D%20limit%2010&endpoint=https%3A%2F%2Frcf-sparql.demo.logilab.fr%2Fquery&requestMethod=POST&tabTitle=Query&headers=%7B%7D&contentTypeConstruct=application%2Fn-triples%2C*%2F*%3Bq%3D0.9&contentTypeSelect=application%2Fsparql-results%2Bjson%2C*%2F*%3Bq%3D0.9&union-default-graph=true&outputFormat=table) : 

[When there is something to evaluate, it hangs](https://rcf-sparql.demo.logilab.fr/#query=PREFIX%20rcf%3A%20%3Chttp%3A%2F%2Frcf.logilab.fr%2F%3E%0APREFIX%20rdfs%3A%20%3Chttp%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23%3E%0APREFIX%20xsd%3A%20%3Chttp%3A%2F%2Fwww.w3.org%2F2001%2FXMLSchema%23%3E%0ASELECT%20DISTINCT%20*%0A%20%20%20%20%20%20WHERE%20%7B%0A%20%20%09%09%3FexpenseType%20rdfs%3AsubClassOf%20rcf%3AD%C3%A9penseImprim%C3%A9e.%0A%20%20%20%20%20%20%20%20%3FexpenseType%20rdfs%3Alabel%20%3Flabel.%0A%20%20%09%09%3Fexpense%20a%20%3FexpenseType.%0A%20%20%09%09%3Fjournee%20rcf%3AaPourDepense%20%3Fexpense.%0A%20%20%09%09%3Fjournee%20rcf%3AaPourDate%20%3Fdate.%20%0A%20%20%09%09FILTER%20(%3Fdate%20%3D%20%221704-09-07%22%5E%5Exsd%3Adate)%0A%20%20%20%20%20%20%7D%20limit%2010&endpoint=https%3A%2F%2Frcf-sparql.demo.logilab.fr%2Fquery&requestMethod=POST&tabTitle=Query&headers=%7B%7D&contentTypeConstruct=application%2Fn-triples%2C*%2F*%3Bq%3D0.9&contentTypeSelect=application%2Fsparql-results%2Bjson%2C*%2F*%3Bq%3D0.9&union-default-graph=true&outputFormat=table)

When ran from the client we get `error at 18:3: expected <unreported>` in the Response.

Those queries return the expected data when run against a fuseki server.

Do you have an idea how to troubleshoot this issue ? 

Thanks a lot for your work !

**Comment by Tpt** [comment]
Hi! Thank you for trying Oxigraph! 

The error always happens after 50s + some milliseconds. Are you using a reverse proxy with a timeout?

It's also strange that the query with the filter is way slower than the query without. There might be a bug behind it. Is there anything on the stdout or stderr output of the `oxigraph_server` process?

**Comment by Murloc6** [comment]
Hi, I am working with @YoelisA about this experimentation.
I looked at the logs and nothing is logged when the timeout occurs.
We are deploying this oxigraph server thought a k8s cluster, then the Ingress is used as a reverse proxy. We can extend the timeout limitation, but the problem is mainly understanding why the second query is this long. 

Thank you for your answer :)

And many thanks for your work on Oxigraph, this is such a nice Sparql Endpoint. Really promising and easy to deploy.

**Comment by Murloc6** [comment]
Definitively, :) I can share an RDF dump of the dataset if it is easier for you.

The explain feature is a so interesting to debug SPARQL queries and see what take long time. For now, we just remove triples until the answer response time is acceptable.

**Comment by Tpt** [comment]
Thank you! I have be able to 700Mo of data directly from the SPARQL endpoint. I hope it's going to be enough to see the bottleneck.

The query plans looks quite similar and nothing obvious to me gets out of them. I'm going to profile the queries and see if I find something with it.

<details>
  <summary>Hardly readable Oxigraph query plans</summary>
  
With filter:
```
Limit { child: HashDeduplicate { child: Project { child: ForLoopJoin { left: ForLoopJoin { left: ForLoopJoin { left: ForLoopJoin { left: QuadPattern { subject: Variable(2), predicate: Constant(NamedNode { iri_id: StrHash { hash: 278574393550247760948885083940562274005 } }), object: Constant(NamedNode { iri_id: StrHash { hash: 317113104573179185288571040651333758362 } }), graph_name: Constant(DefaultGraph) }, right: QuadPattern { subject: Variable(2), predicate: Constant(NamedNode { iri_id: StrHash { hash: 333397388894407012687794434809570484906 } }), object: Variable(4), graph_name: Constant(DefaultGraph) } }, right: QuadPattern { subject: Variable(1), predicate: Constant(NamedNode { iri_id: StrHash { hash: 81786258954639867309623847018166951672 } }), object: Variable(2), graph_name: Constant(DefaultGraph) } }, right: QuadPattern { subject: Variable(3), predicate: Constant(NamedNode { iri_id: StrHash { hash: 239061072387302322250451682252249895395 } }), object: Variable(1), graph_name: Constant(DefaultGraph) } }, right: Filter { child: QuadPattern { subject: Variable(3), predicate: Constant(NamedNode { iri_id: StrHash { hash: 136075188754995423988762900777789846626 } }), object: Variable(0), graph_name: Constant(DefaultGraph) }, expression: Equal(Variable(0), Constant(DateLiteral(Date { timestamp: Timestamp { value: Decimal { value: 53763004800000000000000000000 }, timezone_offset: None } }))) } }, mapping: [(0, 0), (1, 1), (2, 2), (3, 3), (4, 4)] } }, count: 10 }
```

Without filter:
```
Limit { child: HashDeduplicate { child: Project { child: ForLoopJoin { left: ForLoopJoin { left: ForLoopJoin { left: ForLoopJoin { left: Filter { child: QuadPattern { subject: Variable(2), predicate: Constant(NamedNode { iri_id: StrHash { hash: 278574393550247760948885083940562274005 } }), object: Constant(NamedNode { iri_id: StrHash { hash: 317113104573179185288571040651333758362 } }), graph_name: Constant(DefaultGraph) }, expression: Constant(BooleanLiteral(true)) }, right: QuadPattern { subject: Variable(2), predicate: Constant(NamedNode { iri_id: StrHash { hash: 333397388894407012687794434809570484906 } }), object: Variable(4), graph_name: Constant(DefaultGraph) } }, right: QuadPattern { subject: Variable(1), predicate: Constant(NamedNode { iri_id: StrHash { hash: 81786258954639867309623847018166951672 } }), object: Variable(2), graph_name: Constant(DefaultGraph) } }, right: QuadPattern { subject: Variable(3), predicate: Constant(NamedNode { iri_id: StrHash { hash: 239061072387302322250451682252249895395 } }), object: Variable(1), graph_name: Constant(DefaultGraph) } }, right: QuadPattern { subject: Variable(3), predicate: Constant(NamedNode { iri_id: StrHash { hash: 136075188754995423988762900777789846626 } }), object: Variable(0), graph_name: Constant(DefaultGraph) } }, mapping: [(0, 0), (1, 1), (2, 2), (3, 3), (4, 4)] } }, count: 10 }
```
</details>

**Comment by Tpt** [comment]
Hi! I use Clion built-in profiling tool. But it is not easy to read because of the Vulcano-style entanglement of "next" method. Having runtime reports added to the query explanation tool might be very nice to add (yet an other feature to add...).

I am in vacation until August 1st so I might not have time to look more carefully at this issue until then, sorry.

**Comment by fils** [comment]
Just pinging this issue a bit..   I don't get a hang but FILTER doesn't seem to work in the following

```sparql
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX schema: <https://schema.org/>

SELECT (COUNT(?type) as ?count) ?type
WHERE {
    graph ?g {
        ?s rdf:type ?type
        FILTER (?type IN
                (schema:ResearchProject, schema:Project, schema:Organization, schema:Dataset, schema:CreativeWork, schema:Person, schema:Map, schema:Course, schema:CourseInstance, schema:Event, schema:Vehicle)
        )
    }
}
GROUP BY ?type
ORDER BY DESC(?count)
```

but if the FILTER is removed I can get the counts for all types.

**Comment by Tpt** [comment]
@fils Hi! Thank you for trying Oxigraph and sorry for that! When you state it does not work you mean that the query never finishes?

**Comment by fils** [comment]
@Tpt LOL..  that was a bad issue post by me wasn't it..   

Let me try again..    This is being done with nquads files loaded.  In fact, if you are curious they are public and the URLs for them can be found at [this page](https://github.com/iodepo/odis-arch/tree/schema-dev-df/docs/releases) in the "distributions" section.

 If I do

```sparql
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX schema: <https://schema.org/>

SELECT (COUNT(?type) as ?count) ?type
WHERE {
    graph ?g {
        ?s rdf:type ?type
    }
}
GROUP BY ?type
ORDER BY DESC(?count)
```

I get what I expect.    All the counts return fine..    it is big, will put at the bottom

however, 

```sparql
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX schema: <https://schema.org/>

SELECT (COUNT(?type) as ?count) ?type
WHERE {
    graph ?g {
        ?s rdf:type ?type
        FILTER (?type IN
                (schema:ResearchProject, schema:Project, schema:Organization, schema:Dataset, schema:CreativeWork, schema:Person, schema:Map, schema:Course, schema:CourseInstance, schema:Event, schema:Vehicle)
        )
    }
}
GROUP BY ?type
ORDER BY DESC(?count)
```

gives a raw return of

```json
{
  "head": {
    "vars": [
      "count",
      "type"
    ]
  },
  "results": {
    "bindings": []
  }
}
```


The query without the filter returns

```json
{
  "head": {
    "vars": [
      "count",
      "type"
    ]
  },
  "results": {
    "bindings": [
      {
        "count": {
          "type": "literal",
          "value": "15688",
          "datatype": "http://www.w3.org/2001/XMLSchema#integer"
        },
        "type": {
          "type": "uri",
          "value": "http://schema.org/DataDownload"
        }
      },
      {
        "count": {
          "type": "literal",
          "value": "11522",
          "datatype": "http://www.w3.org/2001/XMLSchema#integer"
        },
        "type": {
          "type": "uri",
          "value": "http://schema.org/DataCatalog"
        }
      },
      {
        "count": {
          "type": "literal",
          "value": "11522",
          "datatype": "http://www.w3.org/2001/XMLSchema#integer"
        },
        "type": {
          "type": "uri",
          "value": "http://schema.org/ContactPoint"
        }
      },
      {
        "count": {
          "type": "literal",
          "value": "11522",
          "datatype": "http://www.w3.org/2001/XMLSchema#integer"
        },
        "type": {
          "type": "uri",
          "value": "http://schema.org/Dataset"
        }
      },
      {
        "count": {
          "type": "literal",
          "value": "11500",
          "datatype": "http://www.w3.org/2001/XMLSchema#integer"
        },
        "type": {
          "type": "uri",
          "value": "http://schema.org/Organization"
        }
      },
      {
        "count": {
          "type": "literal",
          "value": "385",
          "datatype": "http://www.w3.org/2001/XMLSchema#integer"
        },
        "type": {
          "type": "uri",
          "value": "http://schema.org/Thing"
        }
      },
      {
        "count": {
          "type": "literal",
          "value": "233",
          "datatype": "http://www.w3.org/2001/XMLSchema#integer"
        },
        "type": {
          "type": "uri",
          "value": "http://schema.org/GeoShape"
        }
      },
      {
        "count": {
          "type": "literal",
          "value": "233",
          "datatype": "http://www.w3.org/2001/XMLSchema#integer"
        },
        "type": {
          "type": "uri",
          "value": "http://schema.org/Place"
        }
      }
    ]
  }
}
```

**Comment by Tpt** [comment]
@fils  I believe it's because you have `PREFIX schema: <https://schema.org/>` in your query but use `PREFIX schema: <https://schema.org/>` (without `s`) in the data.

**Comment by fils** [comment]
OK..  I just noticed that!   but wait.  I was rather certain my files loaded with https not http.   Let me check that.

**Comment by fils** [comment]
just as a note..  using 
```
 ./oxigraph_server_v0.3.14_x86_64_linux_gnu  serve -l data 
```

**Comment by Tpt** [comment]
> Can that be explained?

I believe it's because you use `PUT` and not `POST` in the  `curl` commands. The `PUT` verb replaces the database content with the request payload and so, discards the already existing content. If the last file you PUT does not contains any https://schema.org/ URL it's not surprising that no results are returned by the query
